helper/identity/templating.go

```diff
@@ -0,0 +1,203 @@
+package identity
+
+import (
+	"errors"
+	"strings"
+)
+
+var (
+	UnbalancedTemplatingCharacterErr = errors.New("unbalanced templating characters")
+	NoEntityAttachedToToken          = errors.New("string contains entity template directives but no entity was provided")
+	NoGroupsAttachedToToken          = errors.New("string contains groups template directives but no groups were provided")
+	TemplateValueNotFound            = errors.New("no value could be found for one of the template directives")
+)
+
+type PopulateStringInput struct {
+	ValidityCheckOnly bool
+	String            string
+	Entity            *Entity
+	Groups            []*Group
+}
+
+func PopulateString(p *PopulateStringInput) (bool, string, error) {
+	if p == nil {
+		return false, "", errors.New("nil input")
+	}
+
+	if p.String == "" {
+		return false, "", nil
+	}
+
+	var subst bool
+	splitStr := strings.Split(p.String, "{{")
+
+	if len(splitStr) >= 1 {
+		if strings.Index(splitStr[0], "}}") != -1 {
+			return false, "", UnbalancedTemplatingCharacterErr
+		}
+		if len(splitStr) == 1 {
+			return false, p.String, nil
+		}
+	}
+
+	var b strings.Builder
+	if !p.ValidityCheckOnly {
+		b.Grow(2 * len(p.String))
+	}
+
+	for i, str := range splitStr {
+		if i == 0 {
+			if !p.ValidityCheckOnly {
+				b.WriteString(str)
+			}
+			continue
+		}
+		splitPiece := strings.Split(str, "}}")
+		switch len(splitPiece) {
+		case 2:
+			subst = true
+			if !p.ValidityCheckOnly {
+				tmplStr, err := performTemplating(strings.TrimSpace(splitPiece[0]), p.Entity, p.Groups)
+				if err != nil {
+					return false, "", err
+				}
+				b.WriteString(tmplStr)
+				b.WriteString(splitPiece[1])
+			}
+		default:
+			return false, "", UnbalancedTemplatingCharacterErr
+		}
+	}
+
+	return subst, b.String(), nil
+}
+
+func performTemplating(input string, entity *Entity, groups []*Group) (string, error) {
+	performAliasTemplating := func(trimmed string, alias *Alias) (string, error) {
+		switch {
+		case trimmed == "id":
+			return alias.ID, nil
+		case trimmed == "name":
+			if alias.Name == "" {
+				return "", TemplateValueNotFound
+			}
+			return alias.Name, nil
+		case strings.HasPrefix(trimmed, "metadata."):
+			val, ok := alias.Metadata[strings.TrimPrefix(trimmed, "metadata.")]
+			if !ok {
+				return "", TemplateValueNotFound
+			}
+			return val, nil
+		}
+
+		return "", TemplateValueNotFound
+	}
+
+	performEntityTemplating := func(trimmed string) (string, error) {
+		switch {
+		case trimmed == "id":
+			return entity.ID, nil
+		case trimmed == "name":
+			if entity.Name == "" {
+				return "", TemplateValueNotFound
+			}
+			return entity.Name, nil
+		case strings.HasPrefix(trimmed, "metadata."):
+			val, ok := entity.Metadata[strings.TrimPrefix(trimmed, "metadata.")]
+			if !ok {
+				return "", TemplateValueNotFound
+			}
+			return val, nil
+		case strings.HasPrefix(trimmed, "aliases."):
+			split := strings.SplitN(strings.TrimPrefix(trimmed, "aliases."), ".", 2)
```

Comment:
It would be probably be good to increase the limit a bit, say to 4 or 5, to catch common user error. You're checking the length anyway, and if they enter more than 2 it would be more useful to return "invalid alias selector" than, for example, "no value could be found for one of the template directives".

Reply:
I don't follow.... That's not the error? And if it's not length two something is very wrong since there is a prefix check just before.